hazelcast/util.py

```diff
@@ -345,3 +346,112 @@ def create_git_info():
 
 def to_list(*args, **kwargs):
     return list(*args, **kwargs)
+
+
+def cmp_to_key(cmp):
+    # 'Convert a cmp= function into a key= function'
+    if cmp is None:
+        return None
+
+    class K:
+        def __init__(self, obj, *args):
+            self.obj = obj
+
+        def __lt__(self, other):
+
+            return cmp(self.obj, other.obj) < 0
+
+        def __gt__(self, other):
+            return cmp(self.obj, other.obj) > 0
+
+        def __eq__(self, other):
+            return cmp(self.obj, other.obj) == 0
+
+        def __le__(self, other):
+            return cmp(self.obj, other.obj) <= 0
+
+        def __ge__(self, other):
+            return cmp(self.obj, other.obj) >= 0
+
+        def __ne__(self, other):
+            return cmp(self.obj, other.obj) != 0
+    return K
+
+
+ITERATION_TYPE = enum(KEY=0, VALUE=1, ENTRY=2)
+
+
+def get_sorted_query_result_set(result_list_future, paging_predicate):
+    """
+    :param result_list_future, a future that returns a list of (K,V) pairs to sort and slice based on paging_predicate
+    attributes.
+    :param paging_predicate
+    :return: list of sorted query results
+    """
+    result_list = result_list_future.result()
+    if len(result_list) == 0:
+        return []
+
+    comparator = paging_predicate.get_comparator()
+    iteration_type = paging_predicate.get_iteration_type()
+
+    result_list.sort(key=_get_comparison_key(comparator, iteration_type))
+
+    nearest_anchor_entry = paging_predicate.get_nearest_anchor_entry()
+    nearest_page = nearest_anchor_entry[0]  # first element in anchor entry pair is index of nearest page.
+    page = paging_predicate.get_page()
+    page_size = paging_predicate.get_page_size()
+    list_size = len(result_list)
+
+    begin = page_size * (page - nearest_page - 1)
+    if begin > list_size:
+        return []
+    end = min(begin + page_size, list_size)
+
+    _set_anchor(result_list, paging_predicate, nearest_page)
+
+    sorted_query_result_set = result_list[begin:end]
```

Comment:
I am not sure I understand how that kind of structure would work since we are returning to the user a page of a specific size, not an entry at a specific index like idx.
I mostly based my approach on other clients like node.js and .NET that have very similar [implementations](https://github.com/hazelcast/hazelcast-csharp-client/blob/3.12.z/Hazelcast.Net/Hazelcast.Util/SortingUtil.cs).

However, in order to avoid double-copying, I suggest an alternate approach that allows us to copy the sublist we are interested in only once (in a manner that is compatible with the iteration type). We would get rid of the `sorted_query_result_set` variable and return the copy directly:
--
```python
    # ...
    _set_anchor(result_list, paging_predicate, nearest_page)

    if iteration_type == ITERATION_TYPE.ENTRY:
        return result_list[begin:end]
    elif iteration_type == ITERATION_TYPE.KEY:
        return [result_list[i][0] for i in range(begin, end)]
    elif iteration_type == ITERATION_TYPE.VALUE:
        return [result_list[i][1] for i in range(begin, end)]
```
--
This should take up very little space since we are only copying a sublist of size page_size.
What do you think about this approach?

Reply:
What you are returning is a list of elements. Users may access the elements at specific indexes or iterate over the list. As long as your custom implementation complies with the list protocol, it works for the user. For example, `ImmutableLazyDataList` defers the deserialization and only does it once when the user requests an element. If the user tries to get the same element, it does not perform the deserialization and returns the deserialized element.

What I proposed is something similar. Let's say you should have returned a list of size 10. But the result_list is size of 20 with 4 and 14 as the begin and end indexes. The data structure could have a return element at 4 + x when the user requests an element at the xth location or returns 14-4 as the length etc.

I am not aware of the page sizes used by the customers. If it is something small, your second implementation could work. Could you check the Hazelcast's documentation or code samples to get an idea about the page sizes?

Reply:
Oh, I see now, that is a smart way to do it. I can try that approach if you think it will be more efficient but I think the page size isn't supposed to be too large.
There is no upper limit for the page size in the documentation, however, the [code samples](https://github.com/hazelcast/hazelcast-code-samples/blob/3.12.z/distributed-map/paging-predicate/src/main/java/PagingPredicateQuery.java) and the [example in the reference manual](https://docs.hazelcast.org/docs/latest/manual/html-single/index.html#filtering-with-paging-predicates) set the page size as something between 3-5 usually. So even if that number is higher, say, 100, I don't think it is costly enough to be worrisome in terms of space complexity. With my second implementation, we return a copy of the portion of the list we are interested in for each query but we don't create extra copies we don't use. So I think it could work.
What is your opinion?

Reply:
Okay then, let's go with your second implementation